edx/edx-enterprise
# [ENT-676] Degreed Integration

# Degreed Integration

This PR adds a new integration channel, namely Degreed.

It necessarily refactors a lot of the existing code infrastructure available for integrated channels. It makes it easier to extend or override base integration channel classes and methods to add new, specific integration channels, like for Degreed.

The structure that this PR aims to give integration channels overall, to be followed by any new integration channel, is the following:

```
Integration Channel
    - Exporters
        - Learner Data
        - Course Metadata
    - Transmitters
        - Learner Data
        - Course Metadata
    - API Client
    - Models
        - Global Integration Channel Configuration
        - Integration Channel Enterprise Customer Configuration
        - Catalog Data Transmission Audit
        - Learner Data Transmission Audit
```

## Exporters

An **exporter** is essentially a **serializer** (I would have gone with this name but we were already using 'Exporter' so I decided to go with it).

An *exporter* will `yield` serialized data given the interface method `export`. We choose to use `yield` as the output of this interface because it allows us to loop over this serialized data at a rate that does not surpass the limit that some integration channel may have for receiving data on the other end.

Exporters may do whatever they must to gather data, including possibly calling other APIs (course API, certificate API, etc.), but make a promise through the `export` method to `yield` some serialized data, to be used by a **transmitter**.

## Transmitters

A **transmitter** transmits exported (read: serialized) data to the integration channel using the API client. It uses the `transmit` interface method which takes a single `payload`, which is an `Exporter` object (so that it can be exported (serialized)). It then delegates the actual transmission of the payload to the API client.

## API Client

The **API Client** for an integration channel is the specialized client that knows how the integration channel expects to be tickled, typically through a RESTful interface.

The *API client* is responsible for, usually, making the RESTful call given a serialized payload that it receives from some *transmitter*. It will handle any authentication details that need to be dealt with for the given integration channel's interface.

## Models

Each integration channel may or may not require certain models. If it does not need to build its own, it may delegate any model requirements to the base integration channel models.

### Global Integration Channel Configuration

A **global integration channel configuration** defines some globally required configuration for a given integration channel, usually, for example, defining API paths, provider IDs, and so on. This configuration is then used throughout the integration channel's code, common to all Enterprise Customers that use this integration channel. 

### Integration Channel Enterprise Customer Configuration

An **integration channel enterprise customer configuration** is a enterprise customer-specific configuration model for the given integration channel; it defines some details that may differ among enterprise customers for the given integration channel.

### Catalog Data Transmission Audit

A **catalog data transmission audit** is a model used to keep track of course metadata that is transmitted to a given integration channel. It may differ according to the integration channel in question, so different integration channels may define their own.

### Learner Data Transmission Audit

A **learner data transmission audit** is a model used to keep track of learner data that is transmitted to a given integration channel. It may differ according to the integration channel in question, so different integration channels may define their own.

**JIRA:** [ENT-676](https://openedx.atlassian.net/browse/ENT-676)

**Merge deadline:** Before end of Sprint W.

**Testing instructions:**

1. Checkout the branch from https://github.com/edx/edx-platform/pull/16457 (`uman/degreed`), and make sure you're installing the `edx-enterprise`'s `uman/degreed` branch as well. Make sure migrations for the Degreed app run successfully, as well as for some of the others.
1. Set up an Enterprise Customer such that it's linked to some Catalog in Discovery, which contains some course(s) that are active, enrollable, and marketable (i.e. has a slug, is published, contains valid seats, has proper enrollment dates, etc..).
1. Setup a Degreed Global Configuration given the Degreed test environment creds (if you don't have them, ask in private).
1. Setup a Degreed Enterprise Customer Configuration, using any of the needed creds.
1. Take note of what service user is on your devstack for communicating with the discovery service (on Docker, it's `discovery_worker` by default).
1. SSH into the business sandbox, and as a normal user (not edxapp) run `sudo tail -f /edx/var/log/lms/edx.log`.

##### Course Metadata

1. Open a new terminal tab, SSH into the business sandbox, and run `./manage.py lms --settings=<environment> transmit_course_metadata --catalog_user <discovery_service_user> --channel DEGREED`.
1. Note that in the tab showing the logs, you get output that corresponds correctly to whatever set of courses you had available (oh, and unless you find a way to filter celery output, you'll see a bunch of other random logs if the sandbox is being used). If you get stuck and it seems like not all courses are being captured from discovery as they should, this is definitely a configuration issue (i.e. with whether courses are properly available, enrollable, etc.).

##### Learner Completion Data

1. SSO into Degreed using the Degreed account you have access to. You should do this by navigating to the course enrollment landing page for the Degreed enterprise you setup. You can use the edX demonstration course.
1. Enroll in the course successfully through the enterprise flow and then complete the course (just pass it at least, i.e. 60%+ grade).
1. Open a new terminal tab, SSH into the business sandbox, and run `./manage.py lms --settings=<environment> transmit_learner_data --api_user staff --channel DEGREED`.
1. Note that in the tab showing the logs, you get output that shows the learner data associated with the SSO details (i.e. email should be the Degreed email; you can see what it is by logging into Degreed, clicking your portrait, and clicking Settings, then going to the 'Email' tab).
1. Look on Degreed's side to see the course marked as completed.
